### ISSUE TITLE
Ill-formatted TypeScript types prevent building (v2.0.9)

### ISSUE THREAD

**Issue by synchronos-t** [recharts/recharts#2513]
- [x] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/fervent-diffie-l8xje)

### Steps to reproduce
1. Use the link above. Do not mind the Codesandbox errors, but instead just use its **Export to ZIP** and download the source.
2. Run `npm install`
3. Run `tsc --build`

### What is expected?
The build should succeed. No output from TypeScript.

### What is actually happening?
TypeScript gives errors:

```
node_modules/recharts/types/polar/Radar.d.ts:66:208 - error TS2304: Cannot find name 'P'.

66         radiusAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGElement> & Omit<React.SVGProps<SVGElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<P, T>> & import("./PolarRadiusAxis").PolarRadiusAxisProps & {
                                                                                                                                                                                                                  ~

node_modules/recharts/types/polar/Radar.d.ts:66:211 - error TS2304: Cannot find name 'T'.

66         radiusAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGElement> & Omit<React.SVGProps<SVGElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<P, T>> & import("./PolarRadiusAxis").PolarRadiusAxisProps & {
                                                                                                                                                                                                                     ~

node_modules/recharts/types/polar/Radar.d.ts:69:215 - error TS2304: Cannot find name 'P'.

69         angleAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGTextElement> & Omit<React.SVGProps<SVGTextElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<P, T>> & import("./PolarAngleAxis").PolarAngleAxisProps & {
                                                                                                                                                                                                                         ~

node_modules/recharts/types/polar/Radar.d.ts:69:218 - error TS2304: Cannot find name 'T'.

69         angleAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGTextElement> & Omit<React.SVGProps<SVGTextElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<P, T>> & import("./PolarAngleAxis").PolarAngleAxisProps & {
```

| Environment | Info |
|---|---|
| Recharts | v2.0.9 |
| React | 17.0.1 |
| System | WSL2/Ubuntu 20.04, Node 14.15.3, NPM 7.7.6, TypeScript 4.2.3 |
| Browser | not applicable (NPM build error) |

---
In essence, TypeScript build will fail, unless I set the `skipLibCheck` to true in tsconfig.json.

2.0.8 doesn't produce this error. (But 2.0.8 uses a version of react-smooth that requires React < 17.)

In 2.0.8, the culprit rows used `Pick<..., (a big list of string literals)>`, and now they use `Omit<..., keyof import("../util/types").DOMAttributesAdaptChildEvent<P, T>>`, which doesn't work as `P` and `T` are not defined.


<!-- generated by reccharts-issue-helper. DO NOT REMOVE -->

**Comment by rvetere** [comment]
The types are so broken, i need to delete "node_modules/recharts/types" manually to get "react-intl-extract-messages" working again... where does the types even come from? I would try to make a fix and create a PR, but i don't even see a "types" folder here in your repository... oO

**Comment by synchronos-t** [comment]
@rvetere Recharts is written in TypeScript. The types are in the source code itself. They're only compiled into /types/*.d.ts files in the package releases.

In this case, e.g. the first error comes from this line: https://github.com/recharts/recharts/blob/e91fad39ec0abeaf7d098b5a26d9603945f68da7/src/polar/Radar.tsx#L95
Which is defined here: https://github.com/recharts/recharts/blob/e91fad39ec0abeaf7d098b5a26d9603945f68da7/src/polar/PolarRadiusAxis.tsx#L27
Which in turn uses this one: https://github.com/recharts/recharts/blob/e91fad39ec0abeaf7d098b5a26d9603945f68da7/src/util/types.ts#L35

Though I can't immediately spot anything wrong with any of those. Somehow they're still compiled into broken .d.ts files.

**Comment by synchronos-t** [comment]
My hunch is that f4d01532251de4494f2cfc22ac7b1313eb394738 is the commit where they broke: [here](https://github.com/recharts/recharts/commit/f4d01532251de4494f2cfc22ac7b1313eb394738#diff-7c3814605c4f31b9c5f5ac643fe72a97797b4f2d13230ea178089879c7e5429c).

**Comment by dzrust** [comment]
As a **Temporary** fix for CI/CD pipelines you can do something like this:
`cat ./recharts-type-fix.txt > ./node_modules/recharts/types/polar/Radar.d.ts`

`recharts-type-fix.txt` duplicates `Radar.d.ts` except `DOMAttributesAdaptChildEvent<P, T>>` becomes `DOMAttributesAdaptChildEvent<any, any>`
It will at least get your build passing until the types generated are fixed.

**Comment by flyon** [comment]
any update on this @xile611 ? Seems like it could be a simple fix for someone knowing the types a bit, and this is currently throwing off the build processes of many people using this otherwise awesome library.

**Comment by achaloux** [comment]
I recently added in recharts and this issue caused my build to fail. I ended up using a yarn patch to apply the temporary fix @dzrust suggested

**Comment by fatso83** [comment]
If you install [`patch-package`](https://www.npmjs.com/package/patch-package) you only need to stuff this into your `patches` folder (typically `pbpaste >  patches/recharts+2.1.0.patch`).
```diff
diff --git a/node_modules/recharts/types/polar/Radar.d.ts b/node_modules/recharts/types/polar/Radar.d.ts
index b36f3cf..b26b692 100644
--- a/node_modules/recharts/types/polar/Radar.d.ts
+++ b/node_modules/recharts/types/polar/Radar.d.ts
@@ -63,10 +63,10 @@ export declare class Radar extends PureComponent<Props, State> {
         animationEasing: string;
     };
     static getComposedData: ({ radiusAxis, angleAxis, displayedData, dataKey, bandSize, }: {
-        radiusAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGElement> & Omit<React.SVGProps<SVGElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<P, T>> & import("./PolarRadiusAxis").PolarRadiusAxisProps & {
+        radiusAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGElement> & Omit<React.SVGProps<SVGElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<any,any>> & import("./PolarRadiusAxis").PolarRadiusAxisProps & {
             scale: (value: any) => number;
         };
-        angleAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGTextElement> & Omit<React.SVGProps<SVGTextElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<P, T>> & import("./PolarAngleAxis").PolarAngleAxisProps & {
+        angleAxis: React.AriaAttributes & import("../util/types").DOMAttributesAdaptChildEvent<any, SVGTextElement> & Omit<React.SVGProps<SVGTextElement>, keyof import("../util/types").DOMAttributesAdaptChildEvent<any, any>> & import("./PolarAngleAxis").PolarAngleAxisProps & {
             scale: (value: any) => number;
         };
         displayedData: any[];
```

**Comment by DoReVo** [comment]
Still an issue in 2.1.2

**Comment by ckifer** [comment]
this _looks_ fixed in 2.1.5+ as of #2675 

if anyone here knows if it is not, please feel free to comment. 

I pulled down the zip, upgraded to 2.1.5 with success except for a warning that recharts was lacking a declaration file, installed the latest 2.2.0 and everything is working. Please comment back if this isn't resolved. Closing for now
